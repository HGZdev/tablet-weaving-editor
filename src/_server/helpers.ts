import jwt from "jsonwebtoken";
import {loadEnv} from "vite";
import {UserRow} from "./graphql/users";

export const getViteConfig = (mode: string | undefined): ImportMetaEnv => {
  if (!mode) throw new Error("getViteConfig: mode is undefined");
  return loadEnv(mode, process.cwd()) as never;
};

const {VITE_JWT_SECRET} = getViteConfig(process.env.NODE_ENV);

export const getUserFromToken = (token: string) => {
  if (!token) return;
  if (!VITE_JWT_SECRET)
    throw new Error("getUserFromToken: secret is undefined");

  try {
    const user = jwt.verify(token, VITE_JWT_SECRET);
    return user;
  } catch (err) {
    console.log(err);
  }
};

export const makeTokenFromUser = (user: UserRow, expiresIn: string = "1hr") => {
  if (!user) throw new Error("makeTokenFromUser: user is undefined");
  if (!VITE_JWT_SECRET)
    throw new Error("makeTokenFromUser: secret is undefined");

  const token = jwt.sign(user, VITE_JWT_SECRET, {
    expiresIn,
  });

  return token;
};
