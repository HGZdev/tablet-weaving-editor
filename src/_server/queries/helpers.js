import {useLazyQuery, useQuery} from 'plugins/apollo'

export const useQueryGen = ({query, variables, options}) => {
	const {data, error, ...rest} = useQuery(query, {variables, ...options})

	const name =
		query?.definitions[0].name.value || (data && Object.keys(data)[0])
	const value = data?.[name]

	if (error) {
		error.graphQLErrors.forEach(({message}) =>
			// eslint-disable-next-line no-console
			console.error('Query failed:', message)
		)
	}

	return {...rest, error, data, value}
}

export const useLazyQueryGen = ({query, variables, options}) => {
	const [load, {data, error, ...rest}] = useLazyQuery(query, {
		variables,
		...options,
	})

	const name =
		query?.definitions[0].name.value || (data && Object.keys(data)[0])
	const value = data?.[name]

	if (error) {
		error.graphQLErrors.forEach(({message}) =>
			// eslint-disable-next-line no-console
			console.error('Query failed:', message)
		)
	}

	return {...rest, error, data, value, load}
}
