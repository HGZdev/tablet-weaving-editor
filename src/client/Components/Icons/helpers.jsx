import React from 'react'
import {Box} from 'Components/Box'
import {styled} from 'Styles'
import {default as Svg} from 'react-inlinesvg'

const ImgWrapBox = styled(Box)`
	svg {
		width: ${p => p.width || p.size};
		height: ${p => p.height || p.size};
	}
`
const SvgBox = styled(Box).attrs({as: Svg})``

export const SvgImage = ({src, size, outerProps, innerProps}) => (
	<ImgWrapBox {...{inline: true, size, ...outerProps}}>
		<SvgBox {...{inline: true, src, ...innerProps}} />
	</ImgWrapBox>
)

export const SvgBgImage = styled(Box).attrs({cursor: true})`
	width: ${p => p.width || p.sizeBg};
	height: ${p => p.height || p.sizeBg};

	border-radius: ${p => p.round && `50%`};

	background-image: url(${p => p.icon});
	background-repeat: no-repeat;
	background-position: center;
	background-size: ${p => p.size};

	transition: 0.5s;
	${p => p.theme.colors}
`
