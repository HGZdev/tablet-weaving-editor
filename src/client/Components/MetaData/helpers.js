const makeImageTags = ({imageUrl, imageMeta}) => [
	...(imageUrl
		? [
				{name: 'image', content: imageUrl},
				{property: 'og:image', content: imageUrl},
		  ]
		: []),
	...(imageMeta
		? [
				{property: 'og:image:width', content: imageMeta.w},
				{property: 'og:image:height', content: imageMeta.h},
		  ]
		: []),
]

const makeVideoTags = ({videoUrl, videoMeta}) => [
	...(videoUrl
		? [
				{name: 'video', content: videoUrl},
				{property: 'og:video:url', content: videoUrl},
		  ]
		: []),
	...(videoMeta
		? [
				{property: 'og:video:type', content: videoMeta.type},
				{property: 'og:video:width', content: videoMeta.width || 200},
				{
					property: 'og:video:height',
					content: videoMeta.height || Math.floor(videoMeta.ratio * 200),
				},
		  ]
		: []),
]

// eslint-disable-next-line complexity
export const makeMetaTagsByTag = ({
	appIcon,
	description,
	imageMeta,
	imageUrl,
	keywords,
	title,
	titleSuffix,
	useTitleSuffix,
	videoMeta,
	videoUrl,
}) => {
	const metaTitle = (useTitleSuffix ? title + titleSuffix : title || '').slice(
		0,
		65
	)
	const metaDesc = (description || '').slice(0, 170)

	return {
		title: [{label: metaTitle}],
		meta: [
			{
				name: 'viewport',
				content: 'width=device-width, initial-scale=1.0, minimum-scale=1',
			},
			{name: 'og:title', property: 'og:title', content: metaTitle},
			{name: 'description', content: metaDesc},
			{
				name: 'og:description',
				property: 'og:description',
				content: metaDesc,
			},
			{name: 'og:type', property: 'og:type', content: 'website'},

			...(keywords ? [{name: 'keywords', content: keywords || ''}] : []),

			...(imageUrl || imageMeta ? makeImageTags({imageUrl, imageMeta}) : []),
			...(videoUrl || videoMeta ? makeVideoTags({videoUrl, videoMeta}) : []),
		],
		link: [
			{rel: 'apple-touch-icon-precomposed', href: appIcon},
			{rel: 'icon', href: appIcon},
		],
	}
}

// in progress
export const makeItemMetaData = value => {
	if (!value) return
	const {
		title,
		fullTitle,
		fullName,
		label,
		description,
		// imageId,
		videoUrl,
	} = value
	// const origin = global.window?.location?.origin || baseUrl

	return {
		title: fullTitle || title || fullName || label,
		useTitleSuffix: true,
		description,
		// imageUrl: imageId && `${origin}${makeUrl({id: imageId})}`,
		videoUrl,
	}
}
