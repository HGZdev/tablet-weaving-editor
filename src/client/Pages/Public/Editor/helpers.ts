// Regular expression for RGB Hex
export const regexRGBHex = /^(#)((?:[\dA-Fa-f]{3}){1,2})$/;

// Array of the hole names
export const holeNames = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("");

// Modulo function to prevent negative results
const modulo = (n: number, m: number) => {
  const remain = n % m;
  return Math.floor(remain >= 0 ? remain : remain + m);
};

// Generate directions matrix
export const genDirsMatrix = (dirsChanges: MakeDirChange[][]) => {
  const matrix: DirChangeTurn[][] = [];

  for (const [i, colArr] of dirsChanges.entries()) {
    if (!colArr) continue;
    let currDir: DirChangeTurn = 1;

    for (const dirChange of colArr) {
      if (!matrix[i]) matrix.push([]);

      if (dirChange) {
        matrix[i].push(0);
        currDir = currDir > 0 ? -1 : 1;
      } else {
        matrix[i].push(currDir || 0);
      }
    }
    currDir = 1;
  }

  return matrix;
};

// Generate square root
export const genSqrt = ({
  tablets,
  row,
  col,
  dirsChanges,
}: {
  tablets: Draft["tablets"];
  row: number;
  col: number;
  dirsChanges: MakeDirChange[][];
}) => {
  const matrix = genDirsMatrix(dirsChanges);
  const tablet = tablets[col];

  // Picks current rotation direction: 1 or -1
  let dirSum = 0;
  for (let i = 0; i <= row; i++) {
    dirSum += dirsChanges[col][i];
  }
  const dir: -1 | 1 = dirSum % 2 === 0 ? 1 : -1;

  const isTurnPoint = matrix[col][row] === 0;

  // The very first line
  if (row === 0) return {bg: tablet[0], dir, isTurnPoint};

  // Picks current thread color on a basis of the turns number (doesn't apply for the first line)
  let turnsSum = -1;
  for (let i = 0; i <= row; i++) {
    turnsSum += matrix[col][i];
  }

  const rest = modulo(turnsSum, tablet.length);
  const bg = tablet[rest];

  return {bg, dir, isTurnPoint};
};

export const genThread = ({
  picks,
  ...props
}: {
  picks: Draft["picks"];
  tablets: Draft["tablets"];
  col: number;
  dirsChanges: MakeDirChange[][];
}) => [...new Array(picks)].map((_x, pick) => genSqrt({row: pick, ...props}));

export const isNumeric = (v: string) =>
  Number.parseInt(v, 10) && !Number.isNaN(v);

export const isJsonString = (str: string) => {
  try {
    JSON.parse(str);
  } catch {
    return false;
  }
  return true;
};

export const downLoadFile = (
  data: string[],
  type: string = "application/json",
  name: string = "newFile"
): void => {
  const blob = new Blob([data.join("\n")], {type: type.toString()});
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.download = name;
  a.href = url;
  document.body.append(a);
  a.click();
  a.remove();
};

export const getSkewLabel: Record<Skew, "S" | "Z"> = {"-1": "Z", "1": "S"};

export const getDirLabel = {"-1": "back", "1": "forward"};
