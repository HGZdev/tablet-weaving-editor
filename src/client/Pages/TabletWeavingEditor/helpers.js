/* eslint-disable unicorn/no-array-reduce */

export const regexRGBHex = /^(#)((?:[\dA-Fa-f]{3}){1,2})$/

export const holeNames = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('')

// prevents returning negative results of modulo
const mod = (n, m) => {
	const remain = n % m
	return Math.floor(remain >= 0 ? remain : remain + m)
}

export const genDirsMatrix = dirsChanges => {
	const matrix = []

	for (const [i, colArr] of dirsChanges.entries()) {
		if (!colArr) continue
		let currDir = 1

		for (const dirChange of colArr) {
			if (!matrix[i]) matrix.push([])

			if (dirChange) {
				matrix[i].push(0)
				currDir = -currDir
			} else {
				matrix[i].push(currDir || 0)
			}
		}
		currDir = 1
	}
	return matrix
}

export const genSqrt = ({tablets, row, col, dirsChanges}) => {
	const matrix = genDirsMatrix(dirsChanges)
	const tablet = tablets[col]

	// picks current rotation direction: 1 or -1
	const dirSum = dirsChanges[col]
		.slice(0, row + 1)
		.reduce((sum, a) => sum + a, 0)
	const dir = !dirSum || mod(dirSum, 2) < 1 ? 1 : -1

	const isTurnPoint = matrix[col][row] === 0

	// the very first line
	if (row === 0) return {bg: tablet[0], dir, isTurnPoint}

	// picks current thread color on a basis of the turns number (doesn't apply for the first line)
	const turnsSum =
		matrix[col].slice(0, row + 1).reduce((sum, turn) => sum + turn, 0) - 1

	const rest = mod(turnsSum, tablet.length)
	const bg = tablet[rest]

	return {bg, dir, isTurnPoint}
}

export const genThread = ({picks, ...props}) =>
	[...new Array(picks)]?.map((_x, pick) => genSqrt({row: pick, ...props}))

export const genDraft = ({tablets, ...props}) =>
	tablets?.map((_x, col) => genThread({col, tablets, ...props}))
