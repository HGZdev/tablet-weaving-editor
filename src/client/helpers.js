import {format, isValid} from 'date-fns'

export const mockFunc = n => n * n

export const dateFormatter = (d, f = 'yyyy-MM-dd') => {
	if (!d) return
	const date = new Date(d)
	if (!isValid(date)) throw new Error(`Date is invalid.`)

	return format(date, f)
}

export const isNumeric = v => Number.parseInt(v, 10) && !Number.isNaN(v)
