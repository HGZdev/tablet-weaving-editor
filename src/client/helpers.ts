import {serialize} from "cookie";

export const setCookie = (
  name: string,
  value: string,
  expiryHours: number
): string => {
  const expires = new Date();
  expires.setTime(expires.getTime() + expiryHours * 60 * 60 * 1000);

  const cookie = serialize(name, value, {
    expires,
    path: "/",
  });

  document.cookie = cookie;
  return cookie;
};

export const destroyCookie = (name: string): string => {
  const expires = new Date();
  expires.setTime(expires.getTime() - 1);

  const cookie = serialize(name, "", {
    expires,
    path: "/",
  });

  document.cookie = cookie;
  return cookie;
};

export const mockFunc = (n: number): number => n * n;

// export const dateFormatter = (
//   d: Date | string,
//   f: string = "yyyy-MM-dd"
// ): string => {
//   if (!d) return "";
//   const date = new Date(d);
//   if (!isValid(date)) throw new Error(`Date is invalid.`);

//   return format(date, f);
// };

export const isNumeric = (v: string | number): boolean => !isNaN(Number(v));

export const isJsonString = (str: string): boolean => {
  try {
    JSON.parse(str);
    return true;
  } catch {
    return false;
  }
};

/**
 * Method is used to download a file.
 * @param data - Array Buffer data
 * @param type - Type of the document.
 * @param name - Name of the document.
 */
export const downLoadFile = (
  data: ArrayBuffer,
  type: string = "application/json",
  name: string = "newFile"
): void => {
  const blob = new Blob([data], {type: type});
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.download = name;
  a.href = url;
  document.body.append(a);
  a.click();
  a.remove();
  // window.open(url)
};
