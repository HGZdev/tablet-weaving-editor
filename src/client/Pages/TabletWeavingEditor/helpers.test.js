import {genDirsMatrix, genSqrt, genThread, genDraft} from './helpers'

const dc1 = [
	[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
	[0, 1, 0, 0, 0, 0, 0, 0, 1, 0],
	[0, 0, 0, 1, 0, 0, 1, 0, 0, 0],
	[0, 1, 0, 0, 0, 1, 0, 0, 0, 0],
]

const matrix1 = [
	[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
	[1, 0, -1, -1, -1, -1, -1, -1, 0, 1],
	[1, 1, 1, 0, -1, -1, 0, 1, 1, 1],
	[1, 0, -1, -1, -1, 0, 1, 1, 1, 1],
]

const t = [
	['lightblue', 'violet', 'orange', 'yellow'], // A
	['lightblue', 'violet', 'orange', 'yellow'], // B
	['lightblue', 'violet', 'orange', 'yellow'], // C
	['lightblue', 'violet', 'orange', 'yellow'], // D
]

test('TAB_01 genDirsMatrix', () => {
	expect(genDirsMatrix(dc1)).toEqual(matrix1)
})

test('TAB_02 genSqrt', () => {
	const tests = [
		[
			{col: 0, row: 0, dirsChanges: dc1, tablets: t},
			{bg: 'lightblue', dir: 1, isTurnPoint: false},
		],
		[
			{col: 0, row: 3, dirsChanges: dc1, tablets: t},
			{bg: 'yellow', dir: 1, isTurnPoint: false},
		],
		[
			{col: 0, row: 6, dirsChanges: dc1, tablets: t},
			{bg: 'orange', dir: 1, isTurnPoint: false},
		],
		[
			{col: 1, row: 9, dirsChanges: dc1, tablets: t},
			{bg: 'yellow', dir: 1, isTurnPoint: false},
		],
		[
			{col: 2, row: 3, dirsChanges: dc1, tablets: t},
			{bg: 'orange', dir: -1, isTurnPoint: true},
		],
		[
			{col: 2, row: 5, dirsChanges: dc1, tablets: t},
			{bg: 'lightblue', dir: -1, isTurnPoint: false},
		],
		[
			{col: 2, row: 6, dirsChanges: dc1, tablets: t},
			{bg: 'lightblue', dir: 1, isTurnPoint: true},
		],
		[
			{col: 3, row: 3, dirsChanges: dc1, tablets: t},
			{bg: 'orange', dir: -1, isTurnPoint: false},
		],
	]

	for (const [input, output] of tests) {
		expect(genSqrt(input)).toEqual(output)
	}
})

test('TAB_03 genThread & genDraft', () => {
	const thread = genThread({col: 2, picks: 10, dirsChanges: dc1, tablets: t})

	expect(thread.map(({bg}) => bg)).toEqual([
		'lightblue',
		'violet',
		'orange',
		'orange',
		'violet',
		'lightblue',
		'lightblue',
		'violet',
		'orange',
		'yellow',
	])

	expect(thread.map(({dir}) => dir)).toEqual([1, 1, 1, -1, -1, -1, 1, 1, 1, 1])

	const draft = genDraft({picks: 10, dirsChanges: dc1, tablets: t})
	expect(draft).toMatchSnapshot()
})
