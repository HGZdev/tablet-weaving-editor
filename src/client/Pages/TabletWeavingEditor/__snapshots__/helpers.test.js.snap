// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TAB_03 genThread & genDraft 1`] = `
Array [
  Array [
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "yellow",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "yellow",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": false,
    },
  ],
  Array [
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": -1,
      "isTurnPoint": true,
    },
    Object {
      "bg": "yellow",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "yellow",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": 1,
      "isTurnPoint": true,
    },
    Object {
      "bg": "yellow",
      "dir": 1,
      "isTurnPoint": false,
    },
  ],
  Array [
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": -1,
      "isTurnPoint": true,
    },
    Object {
      "bg": "violet",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": true,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "yellow",
      "dir": 1,
      "isTurnPoint": false,
    },
  ],
  Array [
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": -1,
      "isTurnPoint": true,
    },
    Object {
      "bg": "yellow",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "orange",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": -1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": true,
    },
    Object {
      "bg": "orange",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "yellow",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "lightblue",
      "dir": 1,
      "isTurnPoint": false,
    },
    Object {
      "bg": "violet",
      "dir": 1,
      "isTurnPoint": false,
    },
  ],
]
`;
